### PR TITLE
Local Spark Provider which supports to submit feature join job in local spark

### LINKED ISSUE #643: Support to Run Feature Join Job in Local Spark Environment
This will help dev to make local test without waiting for cloud spark job finish and enable users to test feature join job before cloud resources are deployed

### PR DESCRIPTION
**Pull request by Yuqing-cat**
## Description

Resolves #643 

## How was this PR tested?

Try `pytest test/test_local_spark_e2e.py`, please remember to set `export SPARK_CONFIG__SPARK_CLUSTER=local`.

* Please notice that current client requires `self.credential = DefaultAzureCredential(
            exclude_interactive_browser_credential=False) if credential is None else credential` to build default registry. 

## Does this PR introduce any user-facing changes?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure to clarify your proposed changes.

### Included in this PR
- [x]  A new `_FeathrDLocalSparkJobLauncher`
- [x] Support `local` as `spark_cluster` after `databricks` and `synapse`
- [x] Test case: `test_local_spark_e2e.py` (a pure local e2e test)
      - non udf feature join
      - udf feature join with wasb observation path
      - feature materialization with Redis online store. (*Spark job has correct logs. However, Redis write seems not working. Will keep investigating.)
- [x] Document: `local-spark-provider.md`  
- [x] `feature_conf/features.conf` is removed so that it will affect test results.

### CODE REVIEW

File: feathr_project/feathr/spark_provider/local_spark_utils/gen_template.sh
Comment (xiaoyongzhu) [inline feathr_project/feathr/spark_provider/local_spark_utils/gen_template.sh:7]: Should we use Jinja to manage the templates, rather than using a plain shell?

File: feathr_project/test/test_local_spark_e2e.py
Comment (xiaoyongzhu) [inline feathr_project/test/test_local_spark_e2e.py:138]: remove this for testing?

**Comment by xiaoyongzhu** [comment]
Thanks! I have minor comments but overall LGTM.

**Comment by blrchen** [comment]
Please investigate if ci test failure is related?

File: feathr_project/feathr/spark_provider/local_spark_utils/gen_template.sh
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/local_spark_utils/gen_template.sh:7]: Not familiar with Jinja, seems a good tool for complex Templates. I could try to adopt it if it's a recommended way.

File: feathr_project/test/test_local_spark_e2e.py
Comment (Yuqing-cat) [reply feathr_project/test/test_local_spark_e2e.py]: removed

**Comment by xiaoyongzhu** [comment]
The scala part should be irrelevant. Sometimes there might be random failures there (not sure why).

But the local testing code is failed. Looks like because the env is forced to be set as synapse or databricks:

https://github.com/linkedin/feathr/blob/main/.github/workflows/pull_request_push_test.yml#L105

so @Yuqing-cat you might want to add a test for local only (i.e. force the env to run only the local test)

**Comment by Yuqing-cat** [comment]
> https://github.com/linkedin/feathr/blob/main/.github/workflows/pull_request_push_test.yml#L105

Is there any best practice for local test? 
A straight forward way would be move `test_local_spark_e2e.py` to anther folder, like `feathr_project/local_test`.
Or I could put `test_local_spark_e2e.py` to `feathr_project/test/local_test` and move all the rest tests into `feathr_project/test/test`. 
However, I'm not sure if it will cause conflict / risks to current code base. 
Any suggestions?

**Comment by windoze** [comment]
> > https://github.com/linkedin/feathr/blob/main/.github/workflows/pull_request_push_test.yml#L105
> 
> Is there any best practice for local test? A straight forward way would be move `test_local_spark_e2e.py` to anther folder, like `feathr_project/local_test`. Or I could put `test_local_spark_e2e.py` to `feathr_project/test/local_test` and move all the rest tests into `feathr_project/test/test`. However, I'm not sure if it will cause conflict / risks to current code base. Any suggestions?

You can just create a test script that uses another `feathr_config.yaml` or set environment to override the spark provider setting, Spark should be able to run within the Docker container.
You may need to reduce the executor number to avoid consuming too many resources, CI env may not like memory hogs.

**Review by windoze — CHANGES_REQUESTED** [review]

File: feathr_project/feathr/spark_provider/local_spark_utils/gen_template.sh
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/local_spark_utils/gen_template.sh]: removed and cal subprocess directly as suggested

File: .github/workflows/pull_request_push_test.yml
Comment (xiaoyongzhu) [inline .github/workflows/pull_request_push_test.yml:226]: we don't have to upload jars to blob in this test.

File: .github/workflows/pull_request_push_test.yml
Comment (xiaoyongzhu) [inline .github/workflows/pull_request_push_test.yml:266]: the comment here needs to be updated

File: .github/workflows/pull_request_push_test.yml
Comment (xiaoyongzhu) [inline .github/workflows/pull_request_push_test.yml:261]: I don't think we need this env `SPARK_CONFIG__AZURE_SYNAPSE__FEATHR_RUNTIME_LOCATION` to run local spark

**Comment by xiaoyongzhu** [comment]
@Yuqing-cat I was looking thru the test failures and feel it's related with this line:

[default_registry_client](https://github.com/linkedin/feathr/blob/main/feathr_project/feathr/client.py#L216)

We should probably make that line optional, if no registry service is set in the YAML config/env config.

**Comment by HanumathRao** [comment]
@Yuqing-cat  Thanks for making the changes. It will be very useful to the newbies to try out e2e test case on the local machine.

I tried to do some local testing based on this branch and am running into an issue (which I think is related to a bug in the feathr code). I am curious if you encountered this issue.

Please find the stack trace below.

`22/09/12 20:07:26 INFO SlidingWindowAggregationJoiner: Selected max window duration PT2160H across all anchors for source abfss://feathrazuretest3fs@feathrazuretest3storage.dfs.core.windows.net/demo_data/green_tripdata_2020-04.csv
Exception in thread "main" java.lang.IllegalArgumentException: Wrong FS: abfss://feathrazuretest3fs@feathrazuretest3storage.dfs.core.windows.net/demo_data/green_tripdata_2020-04.csv/daily, expected: file:///
	at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:807)
	at org.apache.hadoop.fs.RawLocalFileSystem.pathToFile(RawLocalFileSystem.java:105)
	at org.apache.hadoop.fs.RawLocalFileSystem.deprecatedGetFileStatus(RawLocalFileSystem.java:774)
	at org.apache.hadoop.fs.RawLocalFileSystem.getFileLinkStatusInternal(RawLocalFileSystem.java:1100)
	at org.apache.hadoop.fs.RawLocalFileSystem.getFileStatus(RawLocalFileSystem.java:769)
	at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:462)
	at org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1760)
	at com.linkedin.feathr.offline.util.HdfsUtils$.exists(HdfsUtils.scala:453)
	at com.linkedin.feathr.offline.source.pathutil.LocalPathChecker.exists(LocalPathChecker.scala:50)
	at com.linkedin.feathr.offline.source.pathutil.TimeBasedHdfsPathAnalyzer.analyze(TimeBasedHdfsPathAnalyzer.scala:51)
	at com.linkedin.feathr.offline.transformation.AnchorToDataSourceMapper.getWindowAggAnchorDFMapForJoin(AnchorToDataSourceMapper.scala:102)
	at com.linkedin.feathr.offline.swa.SlidingWindowAggregationJoiner.$anonfun$joinWindowAggFeaturesAsDF$8(SlidingWindowAggregationJoiner.scala:142)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:193)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at com.linkedin.feathr.offline.swa.SlidingWindowAggregationJoiner.joinWindowAggFeaturesAsDF(SlidingWindowAggregationJoiner.scala:122)
	at com.linkedin.feathr.offline.join.DataFrameFeatureJoiner.joinSWAFeatures(DataFrameFeatureJoiner.scala:325)
	at com.linkedin.feathr.offline.join.DataFrameFeatureJoiner.joinFeaturesAsDF(DataFrameFeatureJoiner.scala:196)
	at com.linkedin.feathr.offline.client.FeathrClient.joinFeaturesAsDF(FeathrClient.scala:267)
	at com.linkedin.feathr.offline.client.FeathrClient.doJoinObsAndFeatures(FeathrClient.scala:186)
	at com.linkedin.feathr.offline.job.FeatureJoinJob$.getFeathrClientAndJoinFeatures(FeatureJoinJob.scala:141)
	at com.linkedin.feathr.offline.job.FeatureJoinJob$.feathrJoinRun(FeatureJoinJob.scala:193)
	at com.linkedin.feathr.offline.job.FeatureJoinJob$.run(FeatureJoinJob.scala:79)
	at com.linkedin.feathr.offline.job.FeatureJoinJob$.main(FeatureJoinJob.scala:349)
	at com.linkedin.feathr.offline.job.FeatureJoinJob.main(FeatureJoinJob.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:958)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1046)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1055)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
`

**Comment by Yuqing-cat** [comment]
> @Yuqing-cat Thanks for making the changes. It will be very useful to the newbies to try out e2e test case on the local machine.
> 
> I tried to do some local testing based on this branch and am running into an issue (which I think is related to a bug in the feathr code). I am curious if you encountered this issue.
> 
> Please find the stack trace below.
> 
> `22/09/12 20:07:26 INFO SlidingWindowAggregationJoiner: Selected max window duration PT2160H across all anchors for source abfss://feathrazuretest3fs@feathrazuretest3storage.dfs.core.windows.net/demo_data/green_tripdata_2020-04.csv Exception in thread "main" java.lang.IllegalArgumentException: Wrong FS: abfss://feathrazuretest3fs@feathrazuretest3storage.dfs.core.windows.net/demo_data/green_tripdata_2020-04.csv/daily, expected: file:/// at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:807) at org.apache.hadoop.fs.RawLocalFileSystem.pathToFile(RawLocalFileSystem.java:105) at org.apache.hadoop.fs.RawLocalFileSystem.deprecatedGetFileStatus(RawLocalFileSystem.java:774) at org.apache.hadoop.fs.RawLocalFileSystem.getFileLinkStatusInternal(RawLocalFileSystem.java:1100) at org.apache.hadoop.fs.RawLocalFileSystem.getFileStatus(RawLocalFileSystem.java:769) at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:462) at org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1760) at com.linkedin.feathr.offline.util.HdfsUtils$.exists(HdfsUtils.scala:453) at com.linkedin.feathr.offline.source.pathutil.LocalPathChecker.exists(LocalPathChecker.scala:50) at com.linkedin.feathr.offline.source.pathutil.TimeBasedHdfsPathAnalyzer.analyze(TimeBasedHdfsPathAnalyzer.scala:51) at com.linkedin.feathr.offline.transformation.AnchorToDataSourceMapper.getWindowAggAnchorDFMapForJoin(AnchorToDataSourceMapper.scala:102) at com.linkedin.feathr.offline.swa.SlidingWindowAggregationJoiner.$anonfun$joinWindowAggFeaturesAsDF$8(SlidingWindowAggregationJoiner.scala:142) at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293) at scala.collection.immutable.Map$Map1.foreach(Map.scala:193) at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293) at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290) at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108) at com.linkedin.feathr.offline.swa.SlidingWindowAggregationJoiner.joinWindowAggFeaturesAsDF(SlidingWindowAggregationJoiner.scala:122) at com.linkedin.feathr.offline.join.DataFrameFeatureJoiner.joinSWAFeatures(DataFrameFeatureJoiner.scala:325) at com.linkedin.feathr.offline.join.DataFrameFeatureJoiner.joinFeaturesAsDF(DataFrameFeatureJoiner.scala:196) at com.linkedin.feathr.offline.client.FeathrClient.joinFeaturesAsDF(FeathrClient.scala:267) at com.linkedin.feathr.offline.client.FeathrClient.doJoinObsAndFeatures(FeathrClient.scala:186) at com.linkedin.feathr.offline.job.FeatureJoinJob$.getFeathrClientAndJoinFeatures(FeatureJoinJob.scala:141) at com.linkedin.feathr.offline.job.FeatureJoinJob$.feathrJoinRun(FeatureJoinJob.scala:193) at com.linkedin.feathr.offline.job.FeatureJoinJob$.run(FeatureJoinJob.scala:79) at com.linkedin.feathr.offline.job.FeatureJoinJob$.main(FeatureJoinJob.scala:349) at com.linkedin.feathr.offline.job.FeatureJoinJob.main(FeatureJoinJob.scala) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:498) at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52) at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:958) at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180) at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203) at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90) at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1046) at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1055) at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala) `



> @Yuqing-cat Thanks for making the changes. It will be very useful to the newbies to try out e2e test case on the local machine.
> 
> I tried to do some local testing based on this branch and am running into an issue (which I think is related to a bug in the feathr code). I am curious if you encountered this issue.
> 
> Please find the stack trace below.
> 
> `22/09/12 20:07:26 INFO SlidingWindowAggregationJoiner: Selected max window duration PT2160H across all anchors for source abfss://feathrazuretest3fs@feathrazuretest3storage.dfs.core.windows.net/demo_data/green_tripdata_2020-04.csv Exception in thread "main" java.lang.IllegalArgumentException: Wrong FS: abfss://feathrazuretest3fs@feathrazuretest3storage.dfs.core.windows.net/demo_data/green_tripdata_2020-04.csv/daily, expected: file:/// at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:807) at org.apache.hadoop.fs.RawLocalFileSystem.pathToFile(RawLocalFileSystem.java:105) at org.apache.hadoop.fs.RawLocalFileSystem.deprecatedGetFileStatus(RawLocalFileSystem.java:774) at org.apache.hadoop.fs.RawLocalFileSystem.getFileLinkStatusInternal(RawLocalFileSystem.java:1100) at org.apache.hadoop.fs.RawLocalFileSystem.getFileStatus(RawLocalFileSystem.java:769) at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:462) at org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1760) at com.linkedin.feathr.offline.util.HdfsUtils$.exists(HdfsUtils.scala:453) at com.linkedin.feathr.offline.source.pathutil.LocalPathChecker.exists(LocalPathChecker.scala:50) at com.linkedin.feathr.offline.source.pathutil.TimeBasedHdfsPathAnalyzer.analyze(TimeBasedHdfsPathAnalyzer.scala:51) at com.linkedin.feathr.offline.transformation.AnchorToDataSourceMapper.getWindowAggAnchorDFMapForJoin(AnchorToDataSourceMapper.scala:102) at com.linkedin.feathr.offline.swa.SlidingWindowAggregationJoiner.$anonfun$joinWindowAggFeaturesAsDF$8(SlidingWindowAggregationJoiner.scala:142) at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293) at scala.collection.immutable.Map$Map1.foreach(Map.scala:193) at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293) at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290) at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108) at com.linkedin.feathr.offline.swa.SlidingWindowAggregationJoiner.joinWindowAggFeaturesAsDF(SlidingWindowAggregationJoiner.scala:122) at com.linkedin.feathr.offline.join.DataFrameFeatureJoiner.joinSWAFeatures(DataFrameFeatureJoiner.scala:325) at com.linkedin.feathr.offline.join.DataFrameFeatureJoiner.joinFeaturesAsDF(DataFrameFeatureJoiner.scala:196) at com.linkedin.feathr.offline.client.FeathrClient.joinFeaturesAsDF(FeathrClient.scala:267) at com.linkedin.feathr.offline.client.FeathrClient.doJoinObsAndFeatures(FeathrClient.scala:186) at com.linkedin.feathr.offline.job.FeatureJoinJob$.getFeathrClientAndJoinFeatures(FeatureJoinJob.scala:141) at com.linkedin.feathr.offline.job.FeatureJoinJob$.feathrJoinRun(FeatureJoinJob.scala:193) at com.linkedin.feathr.offline.job.FeatureJoinJob$.run(FeatureJoinJob.scala:79) at com.linkedin.feathr.offline.job.FeatureJoinJob$.main(FeatureJoinJob.scala:349) at com.linkedin.feathr.offline.job.FeatureJoinJob.main(FeatureJoinJob.scala) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:498) at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52) at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:958) at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180) at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203) at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90) at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1046) at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1055) at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala) `

Hi @HanumathRao , thank you for your feedback. 
Please replace all 'abfss://feathrazuretest3fs@feathrazuretest3storage.dfs.core.windows.net/demo_data/green_tripdata_2020-04.csv' in your config file or python sample code into a local path like './green_tripdata_2020-04.csv' where you should download the sample file into, and see if the issue can be solved.

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (windoze) [inline feathr_project/feathr/spark_provider/_localspark_submission.py:120]: why did you join and then split `cmd`?

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (windoze) [inline feathr_project/feathr/spark_provider/_localspark_submission.py:91]: Break down this line, it's toooooooooo long.

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (windoze) [inline feathr_project/feathr/spark_provider/_localspark_submission.py:201]: Should we put the value of `master` into `feathr_config.yaml`?

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (windoze) [inline feathr_project/feathr/spark_provider/_localspark_submission.py:203]: Clean up the classpath, lots of them (if not all) are from my test script and should not be here.

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (windoze) [inline feathr_project/feathr/spark_provider/_localspark_submission.py:105]: What does this mean?

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (windoze) [inline feathr_project/feathr/spark_provider/_localspark_submission.py:102]: Did you copy the `noop-1.0.jar` into the current directory? I don't find the code does that.

**Review by windoze — CHANGES_REQUESTED** [review]

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/_localspark_submission.py:102]: it's already there: https://github.com/feathr-ai/feathr/blob/main/feathr_project/feathr/spark_provider/noop-1.0.jar

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/_localspark_submission.py]: put them in a separate function with todo items to refine with dynamic modulation in the future.

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/_localspark_submission.py]: Because based on my tests, the argument list cannot be directly used by Popen. The split function will be a more reliable choice.

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/_localspark_submission.py]: updated.

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/_localspark_submission.py]: Agree, and leave default value as 'local[*]'

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/_localspark_submission.py]: updated.

File: feathr_project/feathr/spark_provider/_localspark_submission.py
Comment (Yuqing-cat) [reply feathr_project/feathr/spark_provider/_localspark_submission.py]: cleaned up all of them and the test runs fine.

**Review by windoze — APPROVED** [review]

**Review by xiaoyongzhu — APPROVED** [review]

**Comment by xiaoyongzhu** [comment]
Approved but we also need sure that users can run this locally without setting Azure credentials

**Comment by blrchen** [comment]
@Yuqing-cat can you please take a look at CI failures before merging?

**Comment by Yuqing-cat** [comment]
> @Yuqing-cat can you please take a look at CI failures before merging?
Please check:
https://github.com/feathr-ai/feathr/actions/runs/3087050712/jobs/4992014137
I was intended to create a pure local (no Azure Credential) e2e test but find some credential logic is embedded in spark critical path. Need more effort to decouple them.

**Pull request merged**
